File: Irrgarten/src/irrgarten/Player.java
package irrgarten;
import java.util.ArrayList;

/**
 *
 * @author Sergio Salvador Gil
 */
public class Player extends LabyrinthCharacter{
    
    private static final int MAX_WEAPONS=2;
    private static final int MAX_SHIELDS=3;
    private static final int INITIAL_HEALTH=10;
    private static final int HITS2LOSE=3;
    
    private char number;
    private int consecutiveHits=0;
    
    private WeaponCardDeck weaponCardDeck;
    private ShieldCardDeck shieldCardDeck;
    
    private ArrayList<Weapon> weapons;
    private ArrayList<Shield> shields;
    
    
    public Player(char number, float intelligence, float strength){
        super("Player#"+number,intelligence,strength,INITIAL_HEALTH);
        this.number=number;
        
        this.weaponCardDeck = new WeaponCardDeck();
        this.shieldCardDeck = new ShieldCardDeck();
        
        this.weapons = new ArrayList<>();
        this.shields = new ArrayList<>();
    }
    public Player(Player other){
        super(other);
        
        this.number=other.number;
        this.consecutiveHits = other.consecutiveHits;
        
        this.weaponCardDeck = other.weaponCardDeck;
        this.shieldCardDeck = other.shieldCardDeck;
        
        this.weapons = new ArrayList<>(other.weapons);
        this.shields = new ArrayList<>(other.shields);
    }
    
    public void resurrect(){
        this.setHealth(INITIAL_HEALTH);
        weapons.clear();
        shields.clear();
        resetHits();
    }

    public char getNumber(){
        return number;
    }
    
    public Directions move(Directions direction, ArrayList<Directions> validMoves){
        int size = validMoves.size();
        boolean contained = validMoves.contains(direction);
        
        if (size > 0 && !contained){
            Directions firstElement=validMoves.get(0);
            return firstElement;
        }
        else return direction;
    }
    
    @Override
    public float attack(){
        return (this.getStrength()+this.sumWeapon());
    }
    
    @Override
    public boolean defend(float receivedAttack){
        return this.manageHit(receivedAttack);
    }
    
    public void receiveReward(){
        
        int wRewards=Dice.weaponReward();
        int sRewards=Dice.shieldReward();
        
        for (int i = 0; i < wRewards; i++){
            Weapon wnew=this.newWeapon();
            this.receiveWeapon(wnew);
        }
        for (int i = 0; i < sRewards; i++){
            Shield snew=this.newShield();
            this.receiveShield(snew);
        }
        
        int extraHealth = Dice.healthReward();
        this.setHealth(this.getHealth()+extraHealth);
    }
    
    @Override
    public String toString(){
        String s=super.toString()+"\nConsecutive hits:" + consecutiveHits+"\nweapons:";
        for(int i = 0; i<weapons.size();i++){
            s+=weapons.get(i).toString()+"      ";
        }
        s+="\nshields:";
        for(int i = 0; i<shields.size();i++){
            s+=shields.get(i).toString()+"      ";
        }
        return s;
    }
    
    private void receiveWeapon(Weapon w){
        weapons.removeIf(wi -> wi.discard());
        int size = weapons.size();
        
        if (size<MAX_WEAPONS){
            weapons.add(w);
        }
    }
    
    private void receiveShield(Shield s){
        shields.removeIf(si -> si.discard());
        int size = shields.size();
        
        if (size<MAX_SHIELDS){
            shields.add(s);
        }
    }
    
    private Weapon newWeapon(){
        return weaponCardDeck.nextCard();
    }
    
    private Shield newShield(){
        return shieldCardDeck.nextCard();
    }
    
    protected float sumWeapon(){
        float suma = 0;
        for (int i = 0; i < weapons.size(); i++){
           suma += weapons.get(i).attack();
        }
        return suma;
    }
    
    protected float sumShield(){
        float suma = 0;
        for (int i = 0; i < shields.size(); i++){
           suma += shields.get(i).protect();
        }
        return suma;
    }
    
    protected float defensiveEnergy(){
        return (this.getIntelligence()+this.sumShield());
    }
    
    private boolean manageHit(float receivedAttack){
        float defense = this.defensiveEnergy();
        if (defense < receivedAttack){
            this.gotWounded();
            this.incConsecutiveHits();
        }
        else this.resetHits();
        boolean lose;
        lose = this.dead() || HITS2LOSE==consecutiveHits;
        
        return lose;
    }
    
    private void resetHits(){
        consecutiveHits=0;
    }
    
    private void incConsecutiveHits(){
        consecutiveHits++;
    }
}

File: rb/Irrgarten/Player.rb
#encoding:UTF-8
require_relative'Dice'
require_relative'Weapon'
require_relative'Shield'
require_relative'LabyrinthCharacter'

module Irrgarten
    class Player < LabyrinthCharacter

        attr_reader :number, :consecutive_hits
        
        @@MAX_WEAPONS=2
        @@MAX_SHIELDS=2
        @@INITIAL_HEALTH=10.0
        @@HITS2LOSE=3

        def initialize(number, intelligence, strength)
            super("Player #{number}", intelligence, strength, @@INITIAL_HEALTH)
            @number=number
            @consecutive_hits=0

            @weapons = Array.new
            @shields = Array.new
        end

        def copy(other)
            super
            @number=other.number
            @consecutive_hits=other.consecutive_hits

            @weapons= Array.new
            @shields= Array.new
        end

        def resurrect
            @health=@@INITIAL_HEALTH
            @weapons = []
            @shields = []
            reset_hits
        end

        def get_number
            return @number
        end

        def move(direction, valid_moves)
            size = valid_moves.size
            contained = valid_moves.include?(direction)
            if (size > 0 && !contained)
                first_element=valid_moves[0]
                return first_element
            else return direction
            end
        end

        def attack
            return @strength + sum_weapon
        end

        def defend(received_attack)
            return manage_hit(received_attack)
        end

        def receive_reward
            w_reward = Dice.weapons_reward
            s_reward = Dice.shields_reward

            w_reward.times do 
                wnew = self.new_weapon
                self.receive_weapon(wnew)
            end
            s_reward.times do
                snew = self.new_shield
                self.receive_shield(snew)
            end
            extra_health = Dice.health_reward
            @health += extra_health
        end

        def to_s
            string=super + "\nConsecutive Hits: #{@consecutive_hits}"
            string+="\nweapons:"
            @weapons.each do |wi|
                string+=wi.to_s + "     "
            end
            string+="\nshields:"
            @shields.each do |si|
                string+=si.to_s + "     "
            end
            string
        end

        protected

        def sum_weapon
            suma=0.0

            @weapons.each do |w|
                suma += w.attack
            end
            return suma
        end

        def sum_shield
            suma=0.0

            @shields.each do |s|
                suma += s.protect
            end
            return suma
        end

        def defensive_energy
            return @intelligence+sum_shield
        end

        private


        def receive_weapon(w)
            @weapons.delete_if{|wi| wi.discard}
            if (@weapons.size < @@MAX_WEAPONS)
                @weapons << w
            end
            
        end

        def receive_shield(s)
            @shields.delete_if{|si| si.discard}
            if (@shields.size < @@MAX_SHIELDS)
                @shields << s
            end
        end

        def new_weapon
            return Weapon.new(Dice.weapon_power, Dice.uses_left)
        end

        def new_shield
            return Shield.new(Dice.shield_power, Dice.uses_left)
        end        

        def manage_hit(received_attack)
            defense = self.defensive_energy
            if (defense < received_attack)
                got_wounded
                inc_consecutive_hits
            else
                reset_hits
            end
            if (@consecutive_hits==@@HITS2LOSE || self.dead?)
                lose=true
            else lose = false
            end
            lose
        end

        def reset_hits
            @consecutive_hits=0
        end

        def inc_consecutive_hits
            @consecutive_hits+=1
        end

    end
end